File: sdk/assemblyscript/src/assembly/localtime.ts
/*
 * SPDX-FileCopyrightText: © Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

// @ts-expect-error: decorator
@external("modus_system", "getTimeInZone")
declare function hostGetTimeInZone(tz: string | null): string | null;

// TODO: Find or create a library for AssemblyScript that can handle time zones,
//       and then expose modus_system.getTimeZoneData for that library.

/**
 * Returns the current local time, in ISO 8601 extended format.
 *
 * The time zone is determined in the following order of precedence:
 * - If the X-Time-Zone header is present in the request, the time zone is set to the value of the header.
 * - If the TZ environment variable is set on the host, the time zone is set to the value of the variable.
 * - Otherwise, the time zone is set to the host's local time zone.
 */
export function now(): string {
  const ts = hostGetTimeInZone(null);
  if (ts === null) {
    throw new Error("Failed to get the local time.");
  }
  return ts;
}

/**
 * @deprecated Use `now` instead (lowercase).
 */
export function Now(): string {
  return now();
}

/**
 * Returns the current local time, in ISO 8601 extended format.
 *
 * @param tz - A valid IANA time zone identifier, such as "America/New_York"
 */
export function nowInZone(tz: string): string {
  if (tz === "") {
    throw new Error("A time zone is required.");
  }

  const ts = hostGetTimeInZone(tz);
  if (ts === null) {
    throw new Error("Failed to get the local time.");
  }

  return ts;
}

/**
 * @deprecated Use `nowInZone` instead (lowercase).
 */
export function NowInZone(tz: string): string {
  return nowInZone(tz);
}

/**
 * Returns the local time zone identifier, in IANA format.
 */
export function getTimeZone(): string {
  return process.env.get("TZ");
}

/**
 * @deprecated Use `getTimeZone` instead (lowercase).
 */
export function GetTimeZone(): string {
  return getTimeZone();
}

/**
 * Determines whether the specified time zone is valid.
 * @param tz A time zone identifier, in IANA format.
 * @returns `true` if the time zone is valid; otherwise, `false`.
 */
export function isValidTimeZone(tz: string): bool {
  if (tz === "") {
    return false;
  }
  return hostGetTimeInZone(tz) !== null;
}

/**
 * @deprecated Use `isValidTimeZone` instead (lowercase).
 */
export function IsValidTimeZone(tz: string): bool {
  return isValidTimeZone(tz);
}

File: sdk/go/pkg/localtime/localtime.go
/*
 * Copyright 2024 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

package localtime

import (
	"errors"
	"os"
	"time"
)

// Now returns the current local time.  The time zone is determined in the following order of precedence:
// - If the X-Time-Zone header is present in the request, the time zone is set to the value of the header.
// - If the TZ environment variable is set on the host, the time zone is set to the value of the variable.
// - Otherwise, the time zone is set to the host's local time zone.
func Now() (time.Time, error) {

	// Try to return a time with the proper full time zone.
	if tz := os.Getenv("TZ"); tz != "" {
		if loc, err := hostGetTimeLocation(tz); err == nil {
			return time.Now().In(loc), nil
		}
	}

	// Otherwise, let the host get the local time with the current offset applied.
	return hostGetLocalTime()
}

// NowInZone returns the current time in the given time zone.
// The time zone should be a valid IANA time zone identifier, such as "America/New_York".
func NowInZone(tz string) (time.Time, error) {
	if tz == "" {
		return time.Time{}, errors.New("a time zone is required")
	}

	// Try to return a time with the proper full time zone.
	if loc, err := hostGetTimeLocation(tz); err == nil {
		return time.Now().In(loc), nil
	}

	// Otherwise, let the host get the local time with the current offset applied.
	return hostGetTimeInZone(tz)
}

// GetTimeZone returns the local time zone identifier, in IANA format.
func GetTimeZone() string {
	return os.Getenv("TZ")
}

// GetLocation returns the time.Location for the given time zone.
// The time zone should be a valid IANA time zone identifier, such as "America/New_York".
func GetLocation(tz string) (*time.Location, error) {
	if tz == "" {
		return nil, errors.New("time zone is required")
	}
	return hostGetTimeLocation(tz)
}

// IsValidTimeZone returns true if the given time zone is valid.
func IsValidTimeZone(tz string) bool {
	_, err := hostGetTimeInZone(tz)
	return err == nil
}
